### ISSUE TITLE
redbean: Infinite loop calling `sched_yield` when opening SQLite databases on NixOS/Linux 6.1.58

### ISSUE THREAD

**Issue by s0ph0s-2** [jart/cosmopolitan#962]
When I run the following script using redbean 2.2, I get an infinite loop calling `sched_yield()` on my NixOS machine.

Script:
```lua
s = require "lsqlite3"
s.open_memory()
```

<details>
<summary>strace:</summary>

```
execve("./redbean-2.2.com", ["./redbean-2.2.com", "-i", "test.lua"], 0x7ffcbe75c220 /* 111 vars */) = 0
getpid()                                = 3378249
munmap(0x700000000000, 262144)          = 0
mmap(0x70000003f000, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS|MAP_GROWSDOWN, -1, 0) = 0x70000003f000
mmap(0x700000000000, 4096, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x700000000000
arch_prctl(ARCH_SET_FS, 0x62f180)       = 0
mmap(0x6fc000040000, 65536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x6fc000040000
mmap(0x6fe000040000, 65536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x6fe000040000
mmap(0x100080040000, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x100080040000
sigaltstack({ss_sp=0x100080040000, ss_flags=0, ss_size=262144}, {ss_sp=NULL, ss_flags=SS_DISABLE, ss_size=0}) = 0
rt_sigaction(SIGQUIT, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGFPE, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGILL, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGSEGV, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTRAP, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGABRT, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGBUS, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGURG, {sa_handler=0x5a9f60, sa_mask=~[QUIT ILL TRAP ABRT BUS FPE KILL SEGV STOP URG], sa_flags=SA_RESTORER|SA_ONSTACK|SA_NODEFER|SA_SIGINFO, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
faccessat(AT_FDCWD, "./redbean-2.2.com", F_OK) = 0
getcwd("/home/s0ph0s", 1024)            = 13
openat(AT_FDCWD, "/home/s0ph0s/redbean-2.2.com", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0755, st_size=2347168, ...}) = 0
mmap(0x100080200000, 2347168, PROT_READ, MAP_SHARED|MAP_FIXED, 3, 0) = 0x100080200000
munmap(0x100080200000, 2031616)         = 0
close(3)                                = 0
mmap(0x100080080000, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x100080080000
munmap(0x100080080000, 262144)          = 0
newfstatat(AT_FDCWD, "/home/s0ph0s/redbean-2.2.com.dbg", 0x70000003f620, 0) = -1 ENOENT (No such file or directory)
futex(0x70000003ffac, FUTEX_WAIT_BITSET|FUTEX_CLOCK_REALTIME, 1, NULL, FUTEX_BITSET_MATCH_ANY) = -1 EAGAIN (Resource temporarily unavailable)
sched_getaffinity(0, 128, [0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15]) = 8
getrandom(NULL, 0, 0)                   = 0
rt_sigprocmask(SIG_SETMASK, ~[ABRT KILL STOP], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], ~[ABRT KILL STOP], 8) = 0
prctl(PR_CAPBSET_DROP, CAP_CHOWN)       = -1 EPERM (Operation not permitted)
prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)  = 0
mmap(0x100080080000, 65536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x100080080000
mmap(0x1000800a0000, 131072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x1000800a0000
rt_sigprocmask(SIG_SETMASK, ~[ABRT KILL STOP], [], 8) = 0
mmap(0x6fd000040000, 65536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x6fd000040000
fcntl(2, F_DUPFD_CLOEXEC, 3)            = 3
rt_sigprocmask(SIG_SETMASK, [], ~[ABRT KILL STOP], 8) = 0
close(3)                                = 0
mmap(0x100080090000, 65536, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x100080090000
openat(AT_FDCWD, "/home/s0ph0s/redbean-2.2.com", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0755, st_size=2347168, ...}) = 0
newfstatat(AT_FDCWD, "/home/s0ph0s/redbean-2.2.com", {st_mode=S_IFREG|0755, st_size=2347168, ...}, 0) = 0
mmap(0x100080600000, 2347168, PROT_READ, MAP_PRIVATE|MAP_FIXED, 3, 0) = 0x100080600000
rt_sigprocmask(SIG_SETMASK, ~[ABRT KILL STOP], [], 8) = 0
fcntl(2, F_DUPFD_CLOEXEC, 3)            = 4
rt_sigprocmask(SIG_SETMASK, [], ~[ABRT KILL STOP], 8) = 0
close(4)                                = 0
openat(AT_FDCWD, "test.lua", O_RDONLY)  = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=39, ...}) = 0
readv(4, [{iov_base="s", iov_len=1}, {iov_base=" = require \"lsqlite3\"\ns.open_mem"..., iov_len=4084}], 2) = 39
readv(4, [{iov_base="", iov_len=4058}, {iov_base=NULL, iov_len=0}], 2) = 0
close(4)                                = 0
rt_sigaction(SIGINT, {sa_handler=0x5a9f60, sa_mask=[], sa_flags=SA_RESTORER|SA_RESETHAND, sa_restorer=0x5ae9e0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
...
```
</details>

I'm seeing this on `Linux rhodes 6.1.58 #1-NixOS SMP PREEMPT_DYNAMIC Sun Oct 15 16:32:41 UTC 2023 x86_64 GNU/Linux`. I don't see any similar issue on macOS 14.1.1.

Interestingly, downgrading to redbean 2.0 seems to fix the error.

Do you need any more information from me in order to triage or reproduce this?

**Comment by pkulchenko** [comment]
I don't see this issue on the current master branch (or maybe I run it differently). It looks like `sched_yield()` is looping *after* `SIGINT` happens (so likely after you stop redbean with `Ctrl-C`), no? Do you run it as `redbean -i script.lua` or somehow differently?

**Comment by pkulchenko** [comment]
Can you `print` the value returned `s.open_memory()`? I wonder if the call completes and it hangs after that...

**Comment by s0ph0s-2** [comment]
> Can you `print` the value returned by `s.open_memory()`? I wonder if the call completes and it hangs after that...

Huh, I did that, and it worked fine. And then I tried to reproduce it with the original test script on the same machine and it works fine now? I didn't do anything differently from last night, I guess there was just some transient weirdness on the machine I was using?

I'll close the issue, since I can't reproduce it any more. Thank you for your help troubleshooting @pkulchenko!

**Comment by pkulchenko** [comment]
Good to hear it's working; thank you for the update! Feel free to re-open if you're able to reproduce and have further details.

**Comment by s0ph0s-2** [comment]
Ah, I forgot that the version of redbean I had on the server was the 2.0 one that I had used to confirm that the issue wasn't present in prior versions. I'm able to reproduce it again now that I copied 2.2 over again.

Here's the output when I add a call to `print()` around the `s.open_memory()`:

```
[nix-shell:~]$ ./redbean-2.2.com -i
>: s = require "lsqlite3"
>: print(s.open_memory())
Killed
```

The `Killed` comes from me running `kill -9 $PID` in another tmux window, after about 10 seconds of nothing being printed.

Good news though, I built redbean from 0bffd0943371521c1b93eed55a2dbbcc2e1e6b46 and the issue isn't present! So whatever caused it seems to have been fixed already. If I have time over the US holiday weekend, I'll bisect through the code to see if I can find the commit that's responsible.
